pkg/skaffold/debug/cnb.go

```diff
@@ -19,45 +19,89 @@ package debug
+// After the transformer, updateForCNBImage rewrites the altered command-line back to
+// the form expected by the CNB launcher.
 //
-// The CNB launcher supports three types of launches:
+// CNB images use a special launcher executable as the entrypoint.  This launcher sets up the
+// execution environment as configured by the various buildpacks, and then hands off to the
+// configured launch.  The CNB launcher supports three types of launches:
 //
-//   1. _predefined processes_ are named sets of command+arguments.  There are two types:
-//      - direct: these are passed uninterpreted to os.exec; the command is resolved in PATH
-//        Note that these may actually be configured to execute `/bin/sh -c ...`
-//      - script: the command is treated as a shell script and passed to `sh -c`, the remaining
-//        arguments are added to the shell, and so available as positional arguments.
+//   1. _predefined processes_ are named sets of command+arguments (similar to a container image's
+//      ENTRYPOINT/CMD).  Processes are created by the buildpacks, and oftentimes there is a
+//      buildpack that parses a user's `Procfile` and turns those contents into processes.
+//      There are two types of process definitions:
+//      - *direct*: these are passed uninterpreted to os.exec; the command is resolved in PATH
+//        Note that in practice we see direct-style definitions that execute `/bin/sh -c ...`
+//      - *script*: the command is treated as a shell script and passed to `sh -c`, and any remaining
+//        arguments on the container command-line are added to the shell and so become available
+//        as positional arguments (see https://github.com/buildpacks/lifecycle/issues/218#issuecomment-567091462).
 //        For example: `sh -c 'echo $0 $1 $2 $3' arg0 arg1 arg2 arg3` => `arg0 arg1 arg2 arg3`.
-//        (https://github.com/buildpacks/lifecycle/issues/218#issuecomment-567091462).
-//   2. _direct execs_ where the container's arg[0] == `--` are treated like direct processes
-//   3. _shell scripts_ where the container's arg[0] is the script, and are treated are like indirect processes.
+//   2. _direct execs_: the user can provide a command-line which is treated like a _direct process_.
+//   3. _shell scripts_: the user can provide a shell script as the first argument and any
+//      remaining arguments are available as positional arguments like _script processes_.
 //
-// A key point is that the script-style launches allow support references to environment variables.
-// So we find the command line to be executed, whether that is a script or a direct, and turn it into
-// a normal command-line.  But we also return a rewriter to transform the the command-line back to
-// the original form.
+// Script-style launches support referencing environment variables since they are expanded by the shell.
+//
+// Configuring the launch depends on the CNB Platform API version being used, which is determined by
```

Comment:
Thanks for the details, it was a little bit easier to understand the code.

Reply:
Going through this was painful :-)